pkg/block/mdfetcher.go

```diff
@@ -0,0 +1,103 @@
+package block
+
+import (
+	"context"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+type MetadataFetcher struct {
+	cachedBlockMDs map[ulid.ULID]*metadata.Meta
+	rwLock         *sync.RWMutex
+
+	concurrencyLimit int
+	logger           log.Logger
+	bkt              objstore.Bucket
+}
+
+func NewMetadataFetcher(lg log.Logger, concurrencyLimit int, bkt objstore.Bucket) *MetadataFetcher {
+	return &MetadataFetcher{
+		logger:           lg,
+		concurrencyLimit: concurrencyLimit,
+		rwLock:           &sync.RWMutex{},
+		cachedBlockMDs:   make(map[ulid.ULID]*metadata.Meta),
+		bkt:              bkt,
+	}
+}
+
+// Fetch wil return the mapping from block ID to its metadata that are currently stored in the bucket passed in.
+// If the corresponding value for a block ID key is nil, no valid metadta is retrieved successfully.
```

Comment:
> If the corresponding value for a block ID key is nil, no valid metadta is retrieved successfully.

1. `s/metadta/metadata`
2. Nice idea to delegate the no-meta-json blocks on caller side. I think it makes sense.

However, maybe two maps make sense here then? 
```
(metas map[ulid.ULID]*metadata.Meta, noMeta []ulid.ULID, err  error)
```
This will be easier to use on caller side I think.

Reply:
sorry, not following about point 1. what is `s/metadta/metadata`?

I thought about whether to distinguish those two cases. For the places I have replaced, I haven't found a case that would require these two being distinguished. Did I miss a case that would require this?

Reply:
> sorry, not following about point 1. what is s/metadta/metadata?

https://www.tutorialspoint.com/unix/unix-regular-expressions.htm

(:

Reply:
Hm, what about the case where we remove partially uploaded blocks? (noMeta + delay):

https://github.com/thanos-io/thanos/blob/865d5ec710b5651ead08a7dc73864a6f5dbeeaa9/pkg/compact/compact.go#L197

I think that's when noMeta is useful.

Reply:
I might be wrong here but I think this is the only case that this required and it is covered in the caller's logic - `thanos/pkg/compact/compact.go`. But it does sound like something useful in future. Tempted to add when we are seeing more such usages.